### PR TITLE
General `map-diff` improvements

### PR DESCRIPTION
**Pull request by macabeus**
This PR adds a set of fixes and improvements on `map-diff`:
- fix bug when reading a wml attributes with comments (https://github.com/wesnoth/wesnoth/issues/6862)
- fix bug when reading terrains like `_off^_usr` (https://github.com/wesnoth/wesnoth/issues/6862)
- changes map order on the image output (https://github.com/wesnoth/wesnoth/issues/6862)
- change focus color to pink (https://discord.com/channels/231976805987385345/259976436490829825/997976827131854991)
- add axis and changes notes on the image

You can see a full example on my test repository: https://github.com/macabeus/wesnoth/pull/17

image

image

### CODE REVIEW

File: .github/workflows/map-diff.yml [hunk 1]
```diff
@@ -46,8 +48,12 @@ jobs:
...
 
           ## Write comment body
-          image_link=$(curl -X POST "https://api.imgur.com/3/upload" \
-            -F "image=@\"$diff_image\"" | jq -r ".data.link")
+          curl_result=$(curl -X POST "https://api.imgur.com/3/upload" -F "image=@\"$diff_image\"")
+
+          echo "Imgur result:"
+          echo "$curl_result"
+
+          image_link=$($curl_result | jq -r ".data.link")
```
Comment (soliton-) [inline .github/workflows/map-diff.yml:56]: You probably meant:
```suggestion
          image_link=$(jq -r ".data.link" <<< "$curl_result")
```

File: .github/workflows/map-diff.yml [hunk 2]
```diff
@@ -34,20 +34,26 @@ jobs:
...
+          diff_image=${new_map%.map}.png
+          node build/index.js "../../$map_path" "$new_map" "$diff_image"
 
           ## Compress image
           pngquant --force --output "$diff_image" "$diff_image"
 
           ## Write comment body
-          image_link=$(curl -X POST "https://api.imgur.com/3/upload" \
-            -F "image=@\"$diff_image\"" | jq -r ".data.link")
+          curl_result=$(curl -X POST "https://api.imgur.com/3/upload" -F "image=@\"$diff_image\"")
```
Comment (soliton-) [inline .github/workflows/map-diff.yml:51]: Btw, -F implies POST adding -X is not useful. I would also put all options before positional parameters.

File: utils/wesnoth-map-diff/src/paint.ts [hunk 3]
```diff
@@ -22,71 +31,111 @@ const getTileImageCoordenates = (tileX: number, tileY: number) => {
...
-    if (!miscCode) {
-      return
-    }
-
-    const misctileImage = images.tile[`${baseCode}^${miscCode}`] || images.tile[`^${miscCode}`]
-    output.composite(misctileImage, x + leftPadding, y)
+const producePainters = (output: Jimp, images: ImagesGetter, leftPadding: number) => {
+  const paintTile = (x: number, y: number, baseCode: string, miscCode?: string) => {
+    const tileImage = images.getTile({ baseCode, miscCode })
+    output.composite(tileImage, x + leftPadding, y)
```
Comment (soliton-) [inline utils/wesnoth-map-diff/src/paint.ts:40]: There really should be a check here or in `getTile` and error output if the tile isn't actually found.

**Review by soliton- — CHANGES_REQUESTED** [review]

**Comment by soliton-** [comment]
I wonder if the change notes should rather be in the comment text not the image.

**Comment by soliton-** [comment]
Ah, also something I wanted to mention... it'd be good to add some comments to the WML parser that it is by no means complete or accurate and just needs to parse terrain.cfg successfully. To make it clear that it's not useful for general use currently.

File: utils/wesnoth-map-diff/src/paint.ts [hunk 4]
```diff
@@ -49,12 +58,59 @@ const paint = async (
...
 
   const diffImageWidth = width * 2 + tileImageSize
   const diffImageHeight = height
 
-  new Jimp(diffImageWidth, diffImageHeight, (_err, output) => {
+  new Jimp(diffImageWidth, diffImageHeight, async (_err, output) => {
+    output.opaque()
+
+    const fontPath = path.resolve(__dirname, '../font/DejaVuSandMono.fnt')
+    const font = await Jimp.loadFont(fontPath)
```
[the remark this answers is not in the document]
Comment (macabeus) [reply utils/wesnoth-map-diff/src/paint.ts:72]: Unfortunately, Jimp only supports `fnt`: https://github.com/oliver-moran/jimp/tree/master/packages/jimp#writing-text

**Comment by cooljeanius** [comment]
Any plans to have it handle merge conflicts, too?

**Comment by macabeus** [comment]
@cooljeanius Probably today at night, I'll can to do the requested changes

**Comment by macabeus** [comment]
@soliton- 
> I wonder if the change notes should rather be in the comment text not the image.

I guess that it's better on the image, since probably always is necessary to open the image on a new tab. It'll be easier to read and spot the changes if the text is within the image.

**Comment by macabeus** [comment]
@cooljeanius I don't have plans to add support to merge conflicts =/

**Pull request merged**